pallets/api/src/fungibles/mod.rs

```diff
@@ -177,6 +197,44 @@ pub mod pallet {
 			let spender = T::Lookup::unlookup(spender);
 			AssetsOf::<T>::approve_transfer(origin, id.into(), spender, value)
 		}
+
+		/// Decreases the allowance of a spender.
+		///
+		/// # Parameters
+		/// * `id` - The ID of the asset.
+		/// * `spender` - The account that is allowed to spend the tokens.
+		/// * `value` - The number of tokens to decrease the allowance by.
+		#[pallet::call_index(7)]
+		#[pallet::weight(<T as Config>::WeightInfo::approve(1, 1))]
+		pub fn decrease_allowance(
+			origin: OriginFor<T>,
+			id: AssetIdOf<T>,
+			spender: AccountIdOf<T>,
+			value: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin.clone())
+				.map_err(|e| e.with_weight(T::DbWeight::get().reads(1)))?;
+			let current_allowance = AssetsOf::<T>::allowance(id.clone(), &who, &spender);
+			let spender = T::Lookup::unlookup(spender);
+			let id: AssetIdParameterOf<T> = id.into();
+
+			if value == Zero::zero() {
+				return Ok(Some(Self::weight_approve(0, 0)).into());
+			}
+			// Cancel the aproval and set the new value if `current_allowance` is more than zero
+			AssetsOf::<T>::cancel_approval(origin.clone(), id.clone(), spender.clone())
+				.map_err(|e| e.with_weight(Self::weight_approve(0, 1)))?;
+			if value.is_zero() {
+				return Ok(Some(Self::weight_approve(0, 1)).into());
+			}
+			AssetsOf::<T>::approve_transfer(
+				origin,
+				id,
+				spender,
+				current_allowance.saturating_sub(value),
+			)?;
+			Ok(Some(Self::weight_approve(1, 1)).into())
```

Comment:
We are checking if `value == 0` twice. This should be able to be done more efficiently

Reply:
Can we refactor this please and finish the PR?

Reply:
I believe what you see is outdated

Reply:
Could you double check this again?